File: src/main/java/com/github/loadtest4j/drivers/gatling/GatlingFactory.java
package com.github.loadtest4j.drivers.gatling;

import com.github.loadtest4j.loadtest4j.Driver;
import com.github.loadtest4j.loadtest4j.DriverFactory;

import java.time.Duration;
import java.util.*;

public class GatlingFactory implements DriverFactory {

    @Override
    public Set<String> getMandatoryProperties() {
        return setOf("url");
    }

    /**
     * Creates a Gatling driver using the following properties.
     *
     * Mandatory properties:
     *
     * - `url`
     *
     * TODO Optional properties:
     *
     */
    @Override
    public Driver create(Map<String, String> properties) {
        final Duration duration = Duration.ofSeconds(Long.valueOf(properties.get("duration")));
        final String url = properties.get("url");

        return new Gatling(url);
    }

    private static Set<String> setOf(String... values) {
        // This utility method can be replaced when Java 9+ is more widely adopted
        final Set<String> internalSet = new HashSet<>(Arrays.asList(values));
        return Collections.unmodifiableSet(internalSet);
    }
}

File: src/main/scala/com/github/loadtest4j/drivers/gatling/GatlingFactory.scala
package com.github.loadtest4j.drivers.gatling

import com.github.loadtest4j.loadtest4j.driver.{Driver, DriverFactory}
import java.util

import scala.collection.JavaConverters
import scala.concurrent.duration._

class GatlingFactory extends DriverFactory {

  override def getMandatoryProperties: util.Set[String] = {
    JavaConverters.setAsJavaSet(Set("duration", "url", "usersPerSecond"))
  }

  /**
    * Creates a Gatling driver using the following properties.
    *
    * Mandatory properties:
    *
    * - `duration`
    * - `url`
    * - `usersPerSecond`
    *
    */
  override def create(properties: util.Map[String, String]): Driver = {
    val duration = properties.get("duration").toLong.seconds
    val url = properties.get("url")
    val usersPerSecond = properties.get("usersPerSecond").toInt

    new Gatling(duration, url, usersPerSecond)
  }

}
